src/supernova/circuit.rs

```diff
@@ -0,0 +1,767 @@
+//! The augmented circuit F' for  SuperNova that includes everything from Nova
+//!   and additionally checks:
+//!    1. Ui and pci are contained in the public output.
+//!    2. Instance is folded into Ui[pci] correctly; just like Nova.
+//!    3. Invokes the function ϕ on input (zi, ωi) to compute pci+1, which represents
+//!    the index of the function Fj currently being run. pci+1 is then sent to the
+//!    next invocation of an augmented circuit (which contains a verifier circuit).
+
+//! There are two Verification Circuits. The primary and the secondary.
+//! Each of them is over a Pasta curve but
+//! only the primary executes the next step of the computation.
+//! We have two running instances. Each circuit takes as input 2 hashes: one for each
+//! of the running instances. Each of these hashes is
+//! H(params = H(shape, ck), i, z0, zi, U). Each circuit folds the last invocation of
+//! the other into the running instance
+
+use crate::{
+  constants::NUM_HASH_BITS,
+  gadgets::{
+    ecc::AllocatedPoint,
+    r1cs::{
+      conditionally_select_alloc_relaxed_r1cs, AllocatedR1CSInstance, AllocatedRelaxedR1CSInstance,
+    },
+    utils::{
+      add_allocated_num, alloc_num_equals, alloc_scalar_as_base, alloc_zero,
+      conditionally_select_vec, le_bits_to_num, scalar_as_base,
+    },
+  },
+  r1cs::{R1CSInstance, RelaxedR1CSInstance},
+  traits::{
+    circuit::SuperNovaStepCircuit, commitment::CommitmentTrait, Group, ROCircuitTrait,
+    ROConstantsCircuit,
+  },
+  Commitment,
+};
+use bellperson::{
+  gadgets::{
+    boolean::{AllocatedBit, Boolean},
+    num::AllocatedNum,
+    Assignment,
+  },
+  Circuit, ConstraintSystem, SynthesisError,
+};
+use ff::Field;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CircuitParams {
+  limb_width: usize,
+  pub n_limbs: usize,
```

Comment:
This `n_limbs` will be used in `RecursiveSNARK.verify` ROCircuit to calculate number of field to RO, instead of hardcod. 
Fixed it to private variable, with a getter `get_n_limbs`